expr_checker.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import ast
+from inspect import getsource
+from textwrap import dedent
+
+
+def ast_default_check_type(method, value):
+    safe_type = (str, int, type(None), type(range(0)))
+
+    if type(value) not in safe_type and not (hasattr(value, "__self__") and type(value.__self__) in safe_type):
+        raise ValueError(f"safe_eval didn't like {value}")
+
+    return value
+
+
+def ast_default_test(func, check_type, *args, **kwargs):
+    for arg in args:
+        check_type("arguments", arg)
+
+    for arg in kwargs.values():
+        check_type("arguments", arg)
+
+    if func.__name__ == "<lambda>":
+        return check_type("returned", func(*args))
+    else:
+        return check_type("returned", func(*args, **kwargs))
+
+
+class NodeChecker(ast.NodeTransformer):
+    def __init__(self, check_fn, allow_function_calls, check_type_fn, get_attr):
+        self.check_fn = check_fn.__name__
+        self.check_type_fn = check_type_fn.__name__
+        self.fncall = allow_function_calls
+        self.getattr = get_attr.__name__
+
+        self.reserved_name = [self.check_fn, self.check_type_fn, self.getattr]
+
+        super().__init__()
+
+    def visit_Call(self, node):
+        self.generic_visit(node)
+
+        if not self.fncall:
+            raise Exception("safe_eval didn't permit you to call any functions")
+
+        if isinstance(node.func, ast.Attribute):
+            return ast.Call(
+                func=ast.Name(self.getattr, ast.Load()),
+                args=[node.func.value, ast.Constant(node.func.attr), node],
+                keywords=[]
+            )
```

Comment:
this should not be here...

I just catched that methods call are not passed to `check_fn`, but they should